[16.0][IMP] cetmix_tower_yaml-manifest_templates_for_YAML_export

Mohajiro opened the pull request:
Allows users to reuse common metadata (license, author, etc.)
in YAML exports by selecting predefined manifest templates
and see this metadata in import wizard.

- New models: yaml.manifest.author, yaml.manifest.tmpl
- Export wizard:
  - Two-page notebook (“Settings” + “Manifest”)
  - M2O template selector to prefill manifest fields
  - Editable `yaml_file_name`
- Import wizard: grouped layout; hides manifest if absent
- Tests covering templates, filename defaults, sanitization, and manifest behavior

Task: 4787

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced management for YAML Manifest Templates and Authors with dedicated UI menus.
  * Enhanced YAML export wizard with manifest template selection, author assignment, license and pricing fields, versioning, and customizable file naming.
  * YAML import wizard now displays manifest metadata when available.

* **Improvements**
  * Redesigned export and import wizard interfaces for clearer manifest handling and preview.
  * Filename sanitization and validation for YAML exports.
  * Added semantic versioning enforcement for manifest versions.

* **Bug Fixes**
  * Removed comment field and related logic from YAML export process.

* **Tests**
  * Added tests for filename generation and sanitization in YAML export wizard.
  * Removed obsolete test for YAML comment insertion.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Mohajiro committed:
[IMP] cetmix_tower_yaml: manifest templates

Allows users to reuse common metadata (license, author, etc.)
in YAML exports by selecting predefined manifest templates
and see this metadata in import wizard.

- New models: yaml.manifest.author, yaml.manifest.tmpl
- Export wizard:
  - Two-page notebook (“Settings” + “Manifest”)
  - M2O template selector to prefill manifest fields
  - Editable `yaml_file_name`
- Import wizard: grouped layout; hides manifest if absent
- Tests covering templates, filename defaults, sanitization, and manifest behavior

Task: 4787

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,61 @@ class CxTowerYamlExportWiz(models.TransientModel):
+            ("EUR", "Euro"),
+            ("USD", "US Dollar"),
+        ],
+        string="Currency",
```
For the moment we want to keep just EUR and Dollar

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,34 @@
+    )
+    author = fields.Char(help="Author information for the manifest")
+    website = fields.Char(help="Website URL for the manifest")
+    license = fields.Selection(
```
You should use a function instead of an inline-coded list in order to make this list easier to extend and modify later in other modules. Like [here](https://github.com/cetmix/cetmix-tower/blob/f0b2febc0ca9ef742fbdcb4679b64cfff0b2bccb/cetmix_tower_server/models/cx_tower_command.py#L87)

Mohajiro replied:
Good proposal, thanks

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,34 @@
+            ("MIT", "MIT"),
+            ("Custom", "Custom"),
+        ],
+        default="AGPL-3",
```
Same for the default value. Use existing code as a reference.

Mohajiro replied:
I deleted default value to remove the default values to avoid automatic pre-filling when no template is selected.

ivs-cetmix commented on cetmix_tower_yaml/security/ir.model.access.csv:
```diff
@@ -3,3 +3,4 @@ access_yaml_export_wizard,Export YAML,model_cx_tower_yaml_export_wiz,group_expor
 access_yaml_export_wizard_download,Export YAML File,model_cx_tower_yaml_export_wiz_download,group_export,1,1,1,1
 access_yaml_import_wizard_upload,Import YAML,model_cx_tower_yaml_import_wiz_upload,group_import,1,1,1,1
 access_yaml_import_wizard,Import YAML,model_cx_tower_yaml_import_wiz,group_import,1,1,1,1
+access_cx_tower_yaml_manifest_tmpl,cx.tower.yaml.manifest.tmpl,model_cx_tower_yaml_manifest_tmpl,group_import,1,1,1,1
```
1. You are adding new fields in the export wizard. Why are you access to the YAML import group?
2. You need to ensure that only the "Root" group members can modify the templates. Other ones should have read-only access.

Mohajiro replied:
Can you check if this approach is better, please?

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,69 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    manifest_author = fields.Char("Author")
+    manifest_website = fields.Char("Website")
+    manifest_license = fields.Selection(
+        [
```
Use the same function that is used in the templates.

Mohajiro replied:
Ready, thx!

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,69 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    def _onchange_manifest_template_id(self):
+        tmpl = self.manifest_template_id
+        if tmpl:
+            self.manifest_name = tmpl.name
```
```python
if tmpl:
    vals = {
    "manifest_author": tmpl.author,
    "manifest_website": tmpl.website,
.........
    }
...
self.update(vals)
```

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,69 @@ class CxTowerYamlExportWiz(models.TransientModel):
+            self.manifest_license = tmpl.license
+            self.manifest_license_text = tmpl.license_text
+            self.manifest_currency = tmpl.currency
+        else:
```
What if someone has changed a template by mistake? Are you sure that we need to remove all the values when a template is changed? Shouldn't we just keep them?

Mohajiro replied:
Good reflection, thank you!

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,69 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        string="Currency",
+    )
+
+    @api.onchange("manifest_template_id")
```
You should use `compute` with `readonly=False` whenever this is possible instead of `onchange`. Pls check the DevMemo.

Mohajiro replied:
Finally it's done and I hope it's good for us

ivs-cetmix reviewed (changes requested):
Looks like the commit/PR description doesn't reflect the subject of the task completely.
We are not adding the templates. We are implementing manifest for the YAML files with templates being just a part of it.

Mohajiro commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,57 @@
+    )
+
+    @api.model
+    def _get_manifest_currency_selection(self):
+        return [
+            ("EUR", "Euro"),
+            ("USD", "US Dollar"),
+        ]
```
We decided to keep the selector methods as instance methods (self).
This is the standard Odoo pattern for dynamic Selection fields

Mohajiro commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,55 @@
+    author = fields.Char(help="Author information for the manifest")
+    website = fields.Char(help="Website URL for the manifest")
+
+    @classmethod
+    def _get_manifest_license_selection(self):
+        return [
+            ("AGPL-3", "AGPL-3"),
+            ("LGPL-3", "LGPL-3"),
+            ("MIT", "MIT"),
+            ("Custom", "Custom"),
+        ]
```
We decided to keep the selector methods as instance methods (self).
This is the standard Odoo pattern for dynamic Selection fields

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,52 +113,72 @@ def onchange_explode_child_records(self):
-            )._prepare_record_for_yaml()
-            if record_yaml_dict:
-                record_list.append(record_yaml_dict)
+            if hasattr(record, "_prepare_record_for_yaml"):
+                record_yaml_dict = record.with_context(
+                    explode_related_record=explode_related_record,
+                    remove_empty_values=remove_empty_values,
+                )._prepare_record_for_yaml()
+                if record_yaml_dict:
+                    record_list.append(record_yaml_dict)
+            else:
+                continue
```
Good proposal, I gonna fix it in the next commit!

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,55 @@
+    )
+
+    @api.model
+    def _get_manifest_currency_selection(self):
```
Docstring is missing

Mohajiro replied:
Ready

ivs-cetmix commented on cetmix_tower_yaml/security/ir.model.access.csv:
```diff
@@ -3,3 +3,6 @@ access_yaml_export_wizard,Export YAML,model_cx_tower_yaml_export_wiz,group_expor
 access_yaml_export_wizard_download,Export YAML File,model_cx_tower_yaml_export_wiz_download,group_export,1,1,1,1
 access_yaml_import_wizard_upload,Import YAML,model_cx_tower_yaml_import_wiz_upload,group_import,1,1,1,1
 access_yaml_import_wizard,Import YAML,model_cx_tower_yaml_import_wiz,group_import,1,1,1,1
+access_cx_tower_yaml_manifest_tmpl,cx.tower.yaml.manifest.tmpl,model_cx_tower_yaml_manifest_tmpl,base.group_system,1,1,1,1
```
How is `base.group_system` related to the YAML manifests?

Mohajiro replied:
Replaced by cetmix_tower_server.group_root

ivs-cetmix commented on cetmix_tower_yaml/security/ir.model.access.csv:
```diff
@@ -3,3 +3,6 @@ access_yaml_export_wizard,Export YAML,model_cx_tower_yaml_export_wiz,group_expor
 access_yaml_import_wizard_upload,Import YAML,model_cx_tower_yaml_import_wiz_upload,group_import,1,1,1,1
 access_yaml_import_wizard,Import YAML,model_cx_tower_yaml_import_wiz,group_import,1,1,1,1
+access_cx_tower_yaml_manifest_tmpl,cx.tower.yaml.manifest.tmpl,model_cx_tower_yaml_manifest_tmpl,base.group_system,1,1,1,1
+access_cx_tower_yaml_manifest_tmpl_read_import,cx.tower.yaml.manifest.tmpl,model_cx_tower_yaml_manifest_tmpl,cetmix_tower_yaml.group_import,1,0,0,0
```
Are those access right really needed for the YAML import group?

Mohajiro replied:
Nope, I was added read rights for export groupe and CRUD for root users, can you check if it good or no, please

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,52 +116,70 @@ def onchange_explode_child_records(self):
-            )._prepare_record_for_yaml()
-            if record_yaml_dict:
-                record_list.append(record_yaml_dict)
+            if hasattr(record, "_prepare_record_for_yaml"):
```
Why do we need this check? And how is it related to the task?

Mohajiro replied:
Thx, it's not related to the task so I was deleted it

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_author.py:
```diff
@@ -0,0 +1,13 @@
+    _description = "YAML Manifest Author"
+    _order = "name"
+
+    name = fields.Char(required=True, translate=True)
```
This field should not be translatable, it may create a mess in the YAML files because we need to store two versions instead of one: original and translated.

Mohajiro replied:
Super, it's ready

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -133,7 +138,7 @@ def onchange_explode_child_records(self):
                 "summary": self.manifest_summary,
                 "description": self.manifest_description,
-                "author": self.manifest_author,
+                "author": ", ".join(self.manifest_author_ids.mapped("name")),
```
Let's use a list here as well. Eg ["Bob", "Memes United"]

Mohajiro replied:
Can U check?

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,78 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        compute="_compute_manifest",
+        readonly=False,
+    )
+    manifest_summary = fields.Text("Summary")
```
Let's use `Char` instead.

Mohajiro replied:
Ready

ivs-cetmix reviewed (changes requested):
Improve the UX. Use groups for field to arrange them.

Group 1: Name, Summary, Authors, Website, Description 
Group 2. Licensing and pricing. No need to show currency if price is 0

Template selection can be put outside the groups. Or find another way for it to look better) 


![image](https://github.com/user-attachments/assets/71985a8a-a287-4492-9e08-9b0c35240355)

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_author.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright (C) 2024 Cetmix OÜ
```
2025. Check all other newly added files too.

Mohajiro replied:
I have just 2 news files, it's done for this files

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,67 @@
+    author_ids = fields.Many2many(
+        "cx.tower.yaml.manifest.author",
+        string="Authors",
+        help="One or several authors, comma-separated in the manifest.",
```
They are not comma separated any longer.

Mohajiro replied:
Thx, ready

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,78 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    manifest_template_id = fields.Many2one(
+        "cx.tower.yaml.manifest.tmpl", string="Manifest Template"
+    )
+    manifest_name = fields.Char(compute="_compute_manifest", readonly=False)
```
You need to add `store=True` to make this field storable. Same for all the other computed fields.

Mohajiro replied:
Done

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,78 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    manifest_template_id = fields.Many2one(
+        "cx.tower.yaml.manifest.tmpl", string="Manifest Template"
+    )
+    manifest_name = fields.Char(compute="_compute_manifest", readonly=False)
```
string="Snippet Name"

Mohajiro replied:
In place

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,67 @@
+        help="List of author names to include in the YAML manifest",
+    )
+
+    @classmethod
```
Do we really need this decorator?

Mohajiro replied:
Nope, I changed it by @api.model

Mohajiro replied:
You’re right, this suffix tweak isn’t required for the manifest feature.
I’ll roll it back to keep the PR focused on the task.

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,52 +133,69 @@ def onchange_explode_child_records(self):
+            file_suffix = (
+                records.reference
+                if len(records) == 1
+                else f"bulk_{len(records)}_records"
```
What is the purpose of this update? It's not related to the task.

Mohajiro replied:
You’re right — this suffix tweak isn’t required for the manifest feature.
I’ll roll it back to keep the PR focused on the task.

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,89 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    manifest_website = fields.Char(
+        compute="_compute_manifest", readonly=False, store=True
+    )
+    manifest_license = fields.Selection(
```
string="License"

Mohajiro replied:
Ready

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,89 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    manifest_author = fields.Char(
+        compute="_compute_manifest", readonly=False, store=True
+    )
+    manifest_website = fields.Char(
```
string="Website"

Mohajiro replied:
here too

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,89 @@ class CxTowerYamlExportWiz(models.TransientModel):
+            "cx.tower.yaml.manifest.tmpl"
+        ]._get_manifest_currency_selection()
+
+    manifest_template_id = fields.Many2one(
```
placeholder="select a pre-defined template"
remove the "string" key, label will be auto generated from the field name.

Mohajiro replied:
It's done

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,89 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    manifest_license_text = fields.Text(
+        compute="_compute_manifest", readonly=False, store=True
+    )
+    manifest_currency = fields.Selection(
```
string="Currency"

Mohajiro replied:
Ready

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +34,89 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        string="Authors",
+        store=True,
+    )
+    manifest_license_text = fields.Text(
```
string="License Text"

Mohajiro replied:
Here too

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,67 @@
+
+    license = fields.Selection(
+        selection="_get_manifest_license_selection",
+        help="License type for the manifest",
```
help="License used for the code snippet"

Mohajiro replied:
Fixed

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,67 @@
+        required=True,
+        help="Name of the manifest template",
+    )
+    author = fields.Char(help="Author information for the manifest")
```
This field is not needed any longer

Mohajiro replied:
Ready

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,43 +143,56 @@ def onchange_explode_child_records(self):
         )
-
-        # Use the YAML export collector for unique records
-        collector = YamlExportCollector()
```
Why are you updating this code? Ensure that you have the latest version of the code from the main branch and rebase your branch?

Mohajiro replied:
Yes, it was a Git issue. Thanks for pointing it out!

ivs-cetmix reviewed (changes requested):
Here is the correct order of the YAML fields:

```yaml 
manifest:
  name: Servers
  summary: Much like servers
  description: Such much like snippet! Can't wait it runs!
  author:
  - Author
  - Author 2
  website: memes.com
  license: Custom
  license_text: Very license!
  price: 200.0
  currency: EUR
```

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,10 +36,94 @@ class CxTowerYamlExportWiz(models.TransientModel):
+    )
+    manifest_summary = fields.Char("Summary")
+    manifest_description = fields.Text("Description")
+    manifest_price = fields.Float("Price")
```
These fields are meant to be filled in manually in the wizard, not inherited from the template—so they’re left as regular (non-computed) fields by design.

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,66 @@
+        ]
+
+    license = fields.Selection(
+        selection="_get_manifest_license_selection",
```
Use the correct naming  for the selection function. Call as a [lambda function](https://github.com/cetmix/cetmix-tower/blob/e618e260337660f4a59759b0524d20858cf1b0f7/cetmix_tower_server/models/cx_tower_server.py#L94)

Mohajiro replied:
Like this?

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -0,0 +1,66 @@
+        ]
+
+    currency = fields.Selection(
+        selection="_get_manifest_currency_selection",
```
Same here

Mohajiro replied:
ready

ivs-cetmix reviewed (changes requested):
Use the [correct file structure](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#models) in the Python files.

ivs-cetmix reviewed (approved):
Code review LGTM

ivs-cetmix commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_template.py:
```diff
@@ -65,3 +72,17 @@ def _check_license_text_for_custom(self):
+        for rec in self:
+            if not semver.match(rec.version):
+                raise ValidationError(
+                    _("Version must be in format Major.Minor.Patch, e.g. 1.2.3")
```
_("Version must be in the Major.Minor.Patch format, e.g. 1.2.3")

Mohajiro replied:
Ready

ivs-cetmix reviewed (changes requested):
I have selected a manifest template and then removed it. 

![image](https://github.com/user-attachments/assets/68c67c4f-e0c1-4fe1-abdf-622099344413)


Here is the YAML file. 

```yaml
cetmix_tower_yaml_version: 1
manifest:
  name: Ppepe
  author:
  - First author
  - Second author
  version: 1.2.4.sdsdsd
  website: site.com
  license: AGPL-3
  currency: EUR
records:
- cetmix_tower_model: command
```

- Version doesn't match the format.
- Currency is present, however the license is GPL. This is not an issue but we should either always show all the fields,  or hide non-custom license fields completely.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -71,13 +176,43 @@ def onchange_explode_child_records(self):
+                ),
+            ]
+
+            manifest = {k: v for k, v in fields_order if v not in (False, None, "", [])}
```
Ready, thx

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -71,13 +176,43 @@ def onchange_explode_child_records(self):
+                (
+                    "currency",
+                    self.manifest_currency
+                    if self.manifest_license == "Custom" and self.manifest_price
+                    else None,
+                ),
```
In this wizard currency is relevant only for paid, Custom-licensed snippets (it’s hidden in the UI for any other licence). Keeping the extra license == "Custom" guard is therefore intentional.

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
Use the [correct file structure](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#models) in the Python files.

Mohajiro replied:
Ready

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,7 +37,123 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        compute="_compute_manifest",
+        readonly=False,
+        store=True,
+        string="Version",
```
`help="Use the Major.Minor.Patch format, e.g. 1.2.3"`

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,72 @@
+                                <field name="manifest_description" />
+                            </group>
+                            <group string="License and pricing">
+                                <field name="manifest_license" />
```
`placeholder="use the Major.Minor.Patch format, e.g. 1.2.3"`

ivs-cetmix reviewed (changes requested):
Do not generate the `manifest:` entry in the YAML file if the "Name" field is empty.
If the "Name" field is not empty following fields become required:
- summary
- authors
- license

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,75 @@
+                        </page>
+                        <page string="Manifest">
+                            <group>
+                                <field
+                                    name="manifest_template_id"
+                                    placeholder="Select a pre-defined template"
+                                />
```
We don’t actually support “archiving” or deactivating manifest templates in this module—every saved template is always considered active and reusable.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,75 @@
+                                    name="manifest_price"
+                                    attrs="{'invisible': [('manifest_license', '!=', 'Custom')]}"
+                                />
+                                <field
+                                    name="manifest_currency"
+                                    attrs="{'invisible': [('manifest_price', '=', 0)]}"
+                                />
```
I prefer this logic, bro.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,75 @@
+                                force_save="1"
+                                attrs="{'invisible': [('preview_code', '=', False)]}"
+                            />
+                        </page>
```
The ACE editor widget is a special full-width control that lives outside our normal two-column grid on purpose, and we need its footer buttons to remain flushed to the right edge. Wrapping it in another group or colspan block would actually make the overall form layout uneven compared to the rest of the OCA-style wizards. We’ll keep those fields ungrouped for consistency.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,33 +162,99 @@ def onchange_explode_child_records(self):
+                else None,
+            ),
+        ]
+        manifest = {k: v for k, v in fields_order if v not in (False, None, "", [])}
```
We deliberately omit price entries when the value is zero—showing a “0” price isn’t meaningful in the manifest, so we’ll leave the filtering as is.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,33 +162,99 @@ def onchange_explode_child_records(self):
-            record_yaml_dict = record.with_context(
+        for rec in records:
+            data = rec.with_context(
                 explode_related_record=explode_related_record,
                 remove_empty_values=remove_empty_values,
                 yaml_collector=collector,
             )._prepare_record_for_yaml()
-            if record_yaml_dict:
-                record_list.append(record_yaml_dict)
+            if data:
+                record_list.append(data)
```
Using rec here follows our existing convention and is clear enough; we don’t need to change it.

ivs-cetmix commented:
> Do not generate the `manifest:` entry in the YAML file if the "Name" field is empty. If the "Name" field is not empty following fields become required:
> 
> * summary
> * authors
> * license

@Mohajiro this is not done yet

Mohajiro commented:
@Ivan I’ve added name validation: if the manifest name is empty, the manifest section will be omitted from the YAML. If the user provides a name, they must also fill in summary, authors, and license. For better UX, I’ve added a placeholder to the manifest name field. Please review and let me know if this works for you.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,78 @@
+                                />
+                            </group>
+                            <group string="Information">
+                                <field
+                                    name="manifest_name"
+                                    placeholder="Leave empty to skip the manifest section"
+                                />
+                                <field name="manifest_summary" />
+                                <field
+                                    name="manifest_author_ids"
+                                    widget="many2many_tags"
+                                />
+                                <field
+                                    name="manifest_version"
+                                    placeholder="use the Major.Minor.Patch format, e.g. 1.2.3"
+                                />
+                                <field name="manifest_website" />
+                                <field name="manifest_description" />
+                            </group>
+                            <group string="License and pricing">
+                                <field name="manifest_license" />
+                                <field
```
We already enforce those rules on the server side in action_generate_yaml_file if manifest_name is set but any of manifest_summary, manifest_author_ids or manifest_license are blank, we raise a ValidationError. Because the manifest section is entirely skipped when manifest_name is empty, adding attrs="{'required': [('manifest_name','!=',False)]}" in the XML isn’t strictly necessary.
Keeping the guard server-side ensures the logic can’t be bypassed, even if someone crafts their own RPC call.

ivs-cetmix commented:
>If the user provides a name, they must also fill in summary, authors, and license. 
@Mohajiro I don't see these changes in the code.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -232,6 +239,21 @@ def action_generate_yaml_file(self):
 
         self.ensure_one()
+
+        if self.manifest_name:
```
> @Mohajiro I don't see these changes in the code.

@ivs-cetmix it's starts there.

ivs-cetmix replied:
You need to add the `attrs required` in the view, this is controlled by the Odoo automatically.

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,33 +162,99 @@ def onchange_explode_child_records(self):
         record_list = []
-        for record in records:
-            record_yaml_dict = record.with_context(
+        for rec in records:
```
Why are you modifying this code?

Mohajiro replied:
Just a minor refactor for readability and to simplify the logic, but the functionality itself is unchanged. If you prefer the original structure, I can restore it.

Mohajiro commented:
@ivs-cetmix  I added attrs in export wizard view so must I delete the code from method or I can keep it?

ivs-cetmix commented:
> @ivs-cetmix I added attrs in export wizard view so must I delete the code from method or I can keep it?

@Mohajiro why do we need that code if those fields are mandatory anyway?

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -56,33 +162,85 @@ def onchange_explode_child_records(self):
+                )
 
-        self.yaml_code = f"{yaml_header}\n{yaml_code}"
+    def _validate_manifest(self):
+        """Logical cross-checks before saving YAML."""
+        if self.manifest_price and not self.manifest_currency:
+            raise ValidationError(_("Currency is required when price is specified"))
+        if self.manifest_license == "Custom" and not self.manifest_license_text:
+            raise ValidationError(_("License text is required for a custom license"))
```
We added  attrs="{'required': [('manifest_name','!=',False)]}" in wizard view so it's okey

Mohajiro commented:
> > @ivs-cetmix I added attrs in export wizard view so must I delete the code from method or I can keep it?
> 
> @Mohajiro why do we need that code if those fields are mandatory anyway?

We don't need it so it was deleted.

Bearnard21 reviewed (approved):
LGTM

ivs-cetmix reviewed (approved):
Code review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.py:
```diff
@@ -35,19 +35,56 @@ class CxTowerYamlImportWiz(models.TransientModel):
+            rec.manifest_description = manifest.get("description")
+
+            authors = manifest.get("author")
+            if isinstance(authors, list | tuple):
```
We can’t revert to (list, tuple) because our pre-commit hook enforces PEP 604 (list | tuple) for Python 3.10+ compatibility.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.py:
```diff
@@ -35,19 +35,56 @@ class CxTowerYamlImportWiz(models.TransientModel):
+    manifest_currency = fields.Char(readonly=True, compute="_compute_secret_list")
+
     @api.depends("yaml_code")
     def _compute_secret_list(self):
```
I added another method for manifest

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,9 +37,114 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        readonly=False,
+        store=True,
+    )
+    manifest_summary = fields.Char("Summary")
```
```
size=160,
help="Short summary that includes core information. 160 symbols max"`
```

Mohajiro replied:
Added

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,85 @@
+                                />
+                                <field
+                                    name="manifest_summary"
+                                    attrs="{'required': [('manifest_name','!=',False)]}"
```
placeholeder=short summary, 160 symbols max

Mohajiro replied:
Same here

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.py:
```diff
@@ -35,19 +35,61 @@ class CxTowerYamlImportWiz(models.TransientModel):
-                record.secret_list = _(
-                    "After import, please check and provide secret values"
-                    " if needed for the following secrets: %(secrets)s",
+        for rec in self:
```
Why are you changing this function? Is it related to the task?

Mohajiro replied:
tried refactoring, but I saw that this approach wasn’t not really good, so I reverted all the variable names.

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -36,9 +37,118 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        compute="_compute_manifest",
+        readonly=False,
+        store=True,
+        string="Snippet Name",
```
help="Leave this field blank if you don't want to create a manifest"

Mohajiro replied:
Done

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,86 @@
+                            <group string="Information">
+                                <field
+                                    name="manifest_name"
+                                    placeholder="Leave empty to skip the manifest section"
```
Leave empty to skip the manifest creation

Mohajiro replied:
Done

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,86 @@
+                                    placeholder="use the Major.Minor.Patch format, e.g. 1.2.3"
+                                />
+                                <field name="manifest_website" />
+                                <field name="manifest_description" />
```
The content of this field can be really large. And may render the form barely readable. 
You should put it outside of group with a placeholder so it would have no label.
Also it's reasonable to use nested groups to optimise the layout, eg:

```xml
<group name="manifest" string="Manifest">
     <group ....>
          <field ....>
     </group>
     <group ....>
          <field ....>
     </group>
</group>
```

Mohajiro replied:
![image](https://github.com/user-attachments/assets/93769d9e-6a00-4908-95f9-924cd14cbf04)
It's better like this?

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.py:
```diff
@@ -35,19 +35,61 @@ class CxTowerYamlImportWiz(models.TransientModel):
     )
     preview_code = fields.Boolean()
 
+    manifest_name = fields.Char(readonly=True, compute="_compute_yaml_data")
```
string="Snippet Name"

Mohajiro replied:
Fixed

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.py:
```diff
@@ -35,19 +35,61 @@ class CxTowerYamlImportWiz(models.TransientModel):
     preview_code = fields.Boolean()
 
+    manifest_name = fields.Char(readonly=True, compute="_compute_yaml_data")
+    manifest_summary = fields.Char(readonly=True, compute="_compute_yaml_data")
```
string="Summary". Same for all other fields, no need for the "Manifest" prefix in labels.

ivs-cetmix replied:
Same problem with formatting here. What if we have a really large text?

Mohajiro replied:
![image](https://github.com/user-attachments/assets/c8e03e5f-69d3-477d-9fdb-b698778669e9)
I make some view's changes can you see with your experience if it's better  or not really?
Thanks you.

ivs-cetmix replied:
Looks really good! One important notice: license text can be also really big. So it's reasonable to put description and license text in notebook tabs, similar to how it's [done](https://apps.odoo.com/apps/modules/18.0/prt_mail_messages_pro) in the Odoo App Store:
![image](https://github.com/user-attachments/assets/84db4ccf-c6b3-4662-bde3-c845a279f5f4)

ivs-cetmix reviewed (changes requested):
You need to improve the layout, make it readable and usable.
Elements must be grouped by logic with their potential data size taken into account.

![image](https://github.com/user-attachments/assets/38395f22-d723-4533-ab82-468c5a6fdf43)

ivs-cetmix reviewed (commented):
We can now [remove](https://github.com/cetmix/cetmix-tower/blob/e3e3578c76aa988b66e3b370f23968a36f7caa6d/cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py#L22) the comment field because it makes no sense any longer with manifests.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -6,32 +6,86 @@
+                                readonly="1"
+                                force_save="1"
+                                attrs="{'invisible': [('preview_code', '=', False)]}"
+                            />
```
Like this it's okey?

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.xml:
```diff
@@ -9,6 +9,68 @@
+                                colspan="4"
+                                placeholder="License text (optional)"
+                                style="height:120px; white-space:pre-wrap; overflow-wrap:anywhere; margin:0 16px 8px 0;"
+                            />
```
Nope, it's o,ey for me like this, i don't want to used static files

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.xml:
```diff
@@ -9,6 +9,68 @@
+                        <field name="manifest_author_string" string="Author" />
+                        <field name="manifest_version" string="Version" />
+                        <field
+                            name="manifest_website"
+                            string="Website"
+                            attrs="{'invisible': [('manifest_website', '=', False)]}"
+                        />
+                    </group>
+                    <group string="License and pricing">
+                        <field name="manifest_license" string="License" />
+                        <field
+                            name="manifest_price"
+                            string="Price"
+                            attrs="{'invisible': [('manifest_price', '=', False)]}"
+                        />
+                        <field
+                            name="manifest_currency"
+                            string="Currency"
+                            attrs="{'invisible': [('manifest_currency', '=', False)]}"
+                        />
```
Our parent <group> only hides the entire Manifest section when none of the core fields are set. We still need per-field invisibility inside that group because each field has its own existence logic

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.xml:
```diff
@@ -9,6 +9,68 @@
+                            string="Description"
+                            attrs="{'invisible':[('manifest_description','=',False)]}"
+                        >
+                            <field
+                                name="manifest_description"
+                                widget="text"
+                                nolabel="1"
+                                colspan="4"
+                            />
+                        </page>
```
It's working in Odoo 16.0

ivs-cetmix reviewed (changes requested):
Let's move code preview into a notebook tab too. It's more convenient. We can also remove the "Preview code" checkbox.
Also I think the "Manifest" label can be removed because it doesn't add any clarity.

![image](https://github.com/user-attachments/assets/1e58a94b-5db7-4610-a3e5-d6bda3fe9482)

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -167,8 +167,11 @@ def onchange_explode_child_records(self):
-                record_list.append(data)
+            if not data:
+                continue
+            if isinstance(data, dict) and list(data.keys()) == ["reference"]:
```
Could you please explain the logic behind this change?

Mohajiro replied:
I added a quick filter so that “empty” entries like {"reference": ...} don’t end up in the top-level list. Those lightweight references are meant only for nested contexts and cause a missing-model error if exported at root. Now only full object dicts make it into record_list, and imports work smoothly.

Mohajiro replied:
I restored the original logic after discussing it with Dinar. He will rework it in a separate task.

ivs-cetmix reviewed (changes requested):
Import wizard looks good now. 
Check the comments and also update the view of the export wizard to improve its UX similar to the import wizard.

<img width="640" height="401" alt="Image" src="https://github.com/user-attachments/assets/dff6fdf2-0916-4e5b-b341-f95d15ea8801" />

ivs-cetmix reviewed (changes requested):
If we have a checkbox that explicitly adds the manifest, then the manifest name must be a mandatory field. And the placeholder doesn't make any sense any longer.
Code preview should be the second tab, while description the first.

![Image](https://github.com/user-attachments/assets/2f1ec0fe-662d-44f0-911c-871f1b515666)

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -9,29 +9,103 @@
+                    />
+                    <group string="Information">
+
+                        <field name="manifest_name" required="1" />
+                        <field
+                            name="manifest_summary"
+                            attrs="{'required': [('manifest_name','!=',False)]}"
+                            placeholder="Short summary, 160 symbols max"
+                        />
```
The real cap is enforced server-side via constraints, so adding a maxlength in the view would be purely cosmetic. The placeholder remains just a helpful hint.

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -9,29 +9,103 @@
+                            force_save="1"
+                            nolabel="1"
+                            colspan="4"
+                        />
```
We purposely keep the ACE editor inside the “Preview code” tab so it only shows when requested and doesn’t bloat the layout of other tabs. Moving it outside would break that UX.

ivs-cetmix reviewed (changes requested):
1. I think we need to add the "Preview Code" checkbox back. In order to hide the code if it's not really needed. Because this this the mode common case.
2. Tab order: Description -> License -> Code preview

![Image](https://github.com/user-attachments/assets/096a3bb7-508b-4a9d-aee3-2ea1df01071e)

Mohajiro commented:
@ivs-cetmix It's done, can you check if it's correct?

ivs-cetmix reviewed (changes requested):
1. Let's arrange those fields in two columns using "group" in order to preserve the vertical space.
2. License text must be required if custom license is selected.

<img width="800" height="605" alt="Image" src="https://github.com/user-attachments/assets/6ed167dc-8537-4ce9-8536-0f4a3ef62e44" />

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -7,31 +7,114 @@
+                    <group string="Information">
+
+                        <field name="manifest_name" required="1" />
+                        <field
+                            name="manifest_summary"
+                            attrs="{'required': [('manifest_name','!=',False)]}"
+                            placeholder="Short summary, 160 symbols max"
+                        />
+                        <field
+                            name="manifest_author_ids"
+                            widget="many2many_tags"
+                            attrs="{'required': [('manifest_name','!=',False)]}"
+                        />
+                        <field
+                            name="manifest_version"
+                            placeholder="Use the Major.Minor.Patch format, e.g. 1.2.3"
+                        />
+                        <field name="manifest_website" />
```
May be with nexy PR, but for the moment I don't want focused on this

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -7,31 +7,115 @@
+                            nolabel="1"
+                            colspan="4"
+                            placeholder="License text"
+                            attrs="{'required': [('manifest_name','!=',False)]}"
```
May be in the next PR

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -7,31 +7,115 @@
+                        string="Description"
+                        attrs="{'invisible': [('add_manifest','=',False)]}"
+                    >
+                        <field
+                            name="manifest_description"
+                            widget="text"
+                            nolabel="1"
+                            colspan="4"
+                            placeholder="Detailed description (optional)"
+                        />
```
Same, may be in the next PR

ivs-cetmix reviewed (changes requested):
I'm getting this error when trying to save a YAML snippet with the LGPL-3 license.

<img width="800" height="494" alt="Image" src="https://github.com/user-attachments/assets/7b5b633d-e42a-471e-8808-0bd594cc79c6" />

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -35,10 +33,119 @@ class CxTowerYamlExportWiz(models.TransientModel):
         " field values from the exported YAML file.",
     )
     preview_code = fields.Boolean()
+    add_manifest = fields.Boolean()
+
```
It's gonna be same, bro

ivs-cetmix reviewed (approved):
Functional review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

ivs-cetmix commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -21,6 +23,13 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        string="YAML File Name",
+        size=255,
+        default=lambda self: self._default_yaml_file_name(),
+        help="Name of downloading file, ex. my_snippet.yaml",
```
`help="Snippet file name without extension, eg 'my_snippet'"`

We have only one option for extension, so we should not allow to edit it. It must be added automatically.

Mohajiro replied:
Okey, placeholder was replaced and logic for suffix to, I modified 'cx_tower_yaml_export_wiz.py' and tests. Can U check if it's okey?

ivs-cetmix reviewed (approved):
Functional review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Mohajiro commented on cetmix_tower_yaml/models/__init__.py:
```diff
@@ -17,3 +17,5 @@
 from . import cx_tower_scheduled_task
 from . import cx_tower_file
 from . import cx_tower_server
+from . import cx_tower_yaml_manifest_template
+from . import cx_tower_yaml_manifest_author
```
It's not really important, so I'll keep it like this

Mohajiro commented on cetmix_tower_yaml/models/cx_tower_yaml_manifest_author.py:
```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
```
We use the same pattern in all .py files

Mohajiro commented on cetmix_tower_yaml/views/cx_tower_yaml_manifest_template_views.xml:
```diff
@@ -0,0 +1,56 @@
+        <field name="name">cx.tower.yaml.manifest.tmpl.tree</field>
+        <field name="model">cx.tower.yaml.manifest.tmpl</field>
+        <field name="arch" type="xml">
+            <tree>
+                <field name="name" />
+                <field name="author_ids" widget="many2many_tags" />
+                <field name="version" />
+                <field name="website" />
+                <field name="license" />
+                <field name="currency" />
+            </tree>
```
Nice-to-have, but can wait. Current list is tiny; tree already lets users sort. We’ll add a dedicated search when the model grows.

GabbasovDinar commented on cetmix_tower_yaml/wizards/cx_tower_yaml_import_wiz.py:
```diff
@@ -53,6 +87,31 @@ def _compute_secret_list(self):
 
+    @api.depends("yaml_code")
+    def _compute_yaml_data(self):
+        for record in self:
+            data = yaml.safe_load(record.yaml_code)
+
+            manifest = data.get("manifest", {}) if isinstance(data, dict) else {}
+            record.manifest_name = manifest.get("name")
+            record.manifest_summary = manifest.get("summary")
+            record.manifest_description = manifest.get("description")
+
+            authors = manifest.get("author")
+            if isinstance(authors, list | tuple):
+                record.manifest_author_string = ", ".join(authors)
+            elif isinstance(authors, str):
+                record.manifest_author_string = authors
+            else:
+                record.manifest_author_string = False
+
+            record.manifest_version = manifest.get("version")
+            record.manifest_website = manifest.get("website")
+            record.manifest_license = manifest.get("license")
+            record.manifest_license_text = manifest.get("license_text")
+            record.manifest_price = manifest.get("price")
+            record.manifest_currency = manifest.get("currency")
```
```python

        for record in self:
            data = yaml.safe_load(record.yaml_code)

            manifest = data.get("manifest", {}) if isinstance(data, dict) else {}
            authors = manifest.get("author")
            if isinstance(authors, list | tuple):
                manifest_author_string = ", ".join(authors)
            elif isinstance(authors, str):
                manifest_author_string = authors
            else:
                manifest_author_string = False

            record.update({
                "manifest_name": manifest.get("name"),
                "manifest_summary": manifest.get("summary"),
                "manifest_description": manifest.get("description"),
                "manifest_author_string": manifest_author_string,
                "manifest_version": manifest.get("version"),
                "manifest_website": manifest.get("website"),
                "manifest_license": manifest.get("license"),
                "manifest_license_text": manifest.get("license_text"),
                "manifest_price": manifest.get("price"),
                "manifest_currency": manifest.get("currency"),
            })
```

GabbasovDinar commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.py:
```diff
@@ -35,10 +42,139 @@ class CxTowerYamlExportWiz(models.TransientModel):
+        return f"{prefix}_{ref}"
+
+    @api.depends("manifest_template_id")
+    def _compute_manifest(self):
+        mapping = {
+            "manifest_name": "name",
+            "manifest_author_ids": "author_ids",
+            "manifest_website": "website",
+            "manifest_license": "license",
+            "manifest_license_text": "license_text",
+            "manifest_currency": "currency",
+            "manifest_version": "version",
+        }
+        for rec in self:
+            tmpl = rec.manifest_template_id
+            if tmpl:
+                for wiz_field, tmpl_field in mapping.items():
+                    if not rec[wiz_field]:
+                        rec[wiz_field] = tmpl[tmpl_field]
```
```suggestion
    def _compute_manifest(self):
        mapping = {
            "manifest_name": "name",
            "manifest_author_ids": "author_ids",
            "manifest_website": "website",
            "manifest_license": "license",
            "manifest_license_text": "license_text",
            "manifest_currency": "currency",
            "manifest_version": "version",
        }
        for rec in self:
            tmpl = rec.manifest_template_id
            if not tmpl:
                continue
            for wiz_field, tmpl_field in mapping.items():
                if not rec[wiz_field]:
                    rec[wiz_field] = tmpl[tmpl_field]
```

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Mohajiro commented on cetmix_tower_yaml/wizards/cx_tower_yaml_export_wiz.xml:
```diff
@@ -7,31 +7,122 @@
-                    <field name="explode_child_records" />
-                    <field name="remove_empty_values" />
-                    <field name="preview_code" />
+                    <field name="yaml_file_name" placeholder="my_snippet.yaml" />
                 </group>
```
We don't need additional code for nothing, so let's keep it

ivs-cetmix commented:
/ocabot merge minor

CetmixGitDrone commented:
What a great day to merge this nice PR. Let's do it!
Prepared branch [16.0-ocabot-merge-pr-322-by-ivs-cetmix-bump-minor](https://github.com/cetmix/cetmix-tower/commits/16.0-ocabot-merge-pr-322-by-ivs-cetmix-bump-minor), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 657ded4e191de7a2a942447e02473e2aac809247. Thanks a lot for contributing to cetmix. ❤️
